src/hooks/progress/use-tutorial-progress.ts

```diff
@@ -0,0 +1,62 @@
+import { useQuery, UseQueryResult } from '@tanstack/react-query'
+import useAuthentication from 'hooks/use-authentication'
+import {
+	getTutorialProgress,
+	GetTutorialProgressResult,
+	progressPercentToLabel,
+} from 'lib/learn-client/api/progress'
+import {
+	TutorialIdCollectionId,
+	TutorialProgressLabel,
+} from 'lib/learn-client/types'
+import { TUTORIAL_PROGRESS_QUERY_ID } from './'
+
+type QueryDataType = TutorialProgressLabel | null
+
+interface UseTutorialProgressResult
+	extends Omit<UseQueryResult<QueryDataType>, 'data'> {
+	tutorialProgressLabel: UseQueryResult<QueryDataType>['data']
+}
+
+/**
+ * Handles checking if there is a progress for the given
+ * `tutorialId` & `collectionId` combination.
+ */
+const useTutorialProgress = ({
+	tutorialId,
+	collectionId,
+}: TutorialIdCollectionId): UseTutorialProgressResult => {
+	// Get the current user's access token
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+
+	// Fetch a progress records by `tutorialId`, then filter by `collectionId`
+	const { data: tutorialProgressLabel, ...restQueryResult } =
+		useQuery<QueryDataType>(
+			[TUTORIAL_PROGRESS_QUERY_ID, tutorialId, collectionId],
+			() =>
+				getTutorialProgress({ accessToken, tutorialId, collectionId }).then(
+					(data: GetTutorialProgressResult) => {
+						/**
+						 * Data may be null, if a progress record does not exist
+						 * for this tutorialId + collectionId combination.
+						 *
+						 * We pass this on to the query consumer, as they may need
+						 * to know whether the record exists or not.
+						 */
+						if (data == null) {
+							return null
+						}
+						return progressPercentToLabel(data.complete_percent)
+					}
+				),
+			{ enabled: !!accessToken }
+		)
```

Comment:
We should store the full, raw data returned from the API in the cache rather than a formatted / pared down version. This enables any part of the app to use the data however it wants on a long term scale, and minimizes the maintenance we have to do in this layer of the data fetching.

Reply:
@ashleemboyer I think a layer of opinion in the query layer here _might_ be valuable. But definitely unsure, would love to get others opinions - cc @BRKalow @kendallstrautman .

The rationale is that we seem to want to consistently refer to `visited`, `in_progress`, and `complete` on the front-end, and that we don't ever refer to actual percentages in the UI. So the `react-query` layer is meant to add that layer of opinion, rather than asking the query consumer to have to parse out the `visited`, `in_progress`, and `complete` status every time they query.

We don't currently seem to have any plans to record or use tutorial progress in a more granular way, though definitely take the point that if we ever do want to change to using full percentages (or rather the specific milestones that the API supports), then we'd need to refactor these hooks (and our display components, of course, though I think we'd need to refactor those either way if we change how we show progress).

Let me know if that seems reasonable - happy to refactor as suggested, with the raw set of records coming from the query and each component that queries doing the parsing, if we'd prefer things that way 👍

Reply:
☝️ At a higher level this will probably also related to imminent `collection` progress work:

In the spike I'd built a similar layer of opinion, where the query layer handled the logic of "count the completed tutorials in this collection" and returned an integer, rather than returning an array of `ApiCollectionTutorialProgress` records. We could instead return the array of `ApiCollectionTutorialProgress` and have the consumer count through the array of records.

Reply:
This would be a good candidate for a [query selector](https://tanstack.com/query/v4/docs/guides/migrating-to-react-query-3#query-data-selectors). This way, the query cache remains purely a response cache and we get future flexibility without burdening the consumer of the hook.

```suggestion
	// Fetch a progress records by `tutorialId`, then filter by `collectionId`
	const { data: tutorialProgressLabel, ...restQueryResult } =
		useQuery<QueryDataType>(
			[TUTORIAL_PROGRESS_QUERY_ID, tutorialId, collectionId],
			() => getTutorialProgress({ accessToken, tutorialId, collectionId }),
			{
				enabled: !!accessToken,
				select: data => {
					  /**
					  * Data may be null, if a progress record does not exist
					  * for this tutorialId + collectionId combination.
					  *
					  * We pass this on to the query consumer, as they may need
					  * to know whether the record exists or not.
					  */
				        if (data == null) {
					        return null
				        }
					return progressPercentToLabel(data.complete_percent)
				}
			}
		)
```

Reply:
I think the cache should hold 0 opinion about the data.